### PR TITLE
0-1 BFS (Zero One Breadth First Search)

### PR DESCRIPTION
**Pull request by RajatShenoi**
### Describe your change:
This algorithm implements breadth first search for even weighted graphs provided the weights are either zero or one.
This is a very useful function that can help many programmers as they need not learn Dijkstra for a graphs weighted 0 or 1.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: graphs/zero_one_bfs.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Each edge of the graph must be either ZERO or ONE
+
+Refer this link: https://cp-algorithms.com/graph/01_bfs.html for more details
+"""
+
+import collections
+import math
+
+
+def zero_one_bfs(graph: dict, number_of_nodes: int, start: int, goal: int) -> None:
```
Comment (CaedenPH) [inline graphs/zero_one_bfs.py:19]: `dict` of what?

File: graphs/zero_one_bfs.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    Shortest weighted path for a ZERO - ONE weighted graph
+
+    The graph (dict) must have all keys from 0 to number_of_nodes - 1
+    The values of each key must be a list of lists of length 2,
+            the first element being the next node and
+            the second elements being the weight of the edge connecting them
+
+    This function implements the zero_one_bfs. Programmers can just add print
+    statements or modify the data_types as per the applications of this.
+
```
Comment (CaedenPH) [inline graphs/zero_one_bfs.py:30]:
```suggestion
```

File: graphs/zero_one_bfs.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    """
+
+    # d[i] is the distance from the node 'start' to the node i
+    distance_from_start = [math.inf for i in range(number_of_nodes)]
+
+    # The distance of a node from itself is ZERO
+    distance_from_start[start] = 0
+
+    queuee: collections.deque = collections.deque([])
```
Comment (CaedenPH) [inline graphs/zero_one_bfs.py:39]:
```suggestion
    queue: collections.deque = collections.deque([])
```

**Comment by RajatShenoi** [comment]
Participating in Hacktoberfest.

File: boolean_algebra/nandgate.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    """
+
+    if input1 == input2 == 1:
+        return 0
+    else:
+        return 1
+
+
+def main() -> None:
+    print("Truth table of NAND Gate:")
```
Comment (realstealthninja) [inline boolean_algebra/nandgate.py:41]: i would've preferred a multi-line string.

File: boolean_algebra/nandgate.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    >>> nand_gate(0.0, 0.0)
+    1
+    >>> nand_gate(0, -7)
+    1
+    """
+
+    if input1 == input2 == 1:
+        return 0
+    else:
+        return 1
```
Comment (realstealthninja) [inline boolean_algebra/nandgate.py:37]:
```suggestion
    return 1
```
using an else here is unnecessary

File: graphs/zero_one_bfs.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                if weight_of_node == 1:
+                    queuee.append(next_node)
+                else:
+                    queuee.appendleft(next_node)
+
+
+def main() -> None:
+    print("This function implements the zero_one_bfs.")
+    print("Programmers can just add print statements or ", end="")
+    print("modify the data_types as per the applications of this.")
```
Comment (realstealthninja) [inline graphs/zero_one_bfs.py:65]: again would prefer a multi-line string

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by yash206 — APPROVED** [review]

File: boolean_algebra/nandgate.py (hunk 4 above)
Comment (RajatShenoi) [reply boolean_algebra/nandgate.py:40]: Please suggest a multi-line version of this.

File: graphs/zero_one_bfs.py (hunk 1 above)
Comment (RajatShenoi) [reply graphs/zero_one_bfs.py:19]: The way the `dict` may be formatted is given in the comments below.

File: graphs/zero_one_bfs.py (hunk 6 above)
Comment (RajatShenoi) [reply graphs/zero_one_bfs.py:65]: I'm not sure what you exactly mean by a multi-line string. It's already divided into three lines.

Comment (realstealthninja) [reply graphs/zero_one_bfs.py:65]:
```py
"""
This is a multi-line string
It's for cases in which 
You have line breaks
And it's automatically
Managed by python
   
Spaces work
"""
```

Comment (realstealthninja) [inline graphs/zero_one_bfs.py:65]:
```suggestion
    print("""
    This function implements the zero_one_bfs
    Programmers can just add print statements or
    modify the data_types as per the applications of this.
    """)
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Using a single multiline string is better than multiple print statements

File: graphs/zero_one_bfs.py (hunk 6 above)
Comment (RajatShenoi) [reply graphs/zero_one_bfs.py:65]: Ah, now I get it. Never knew about this feature before. Will update once I'm back home. Thanks!

**Comment by cclauss** [comment]
No doctests.  Closing.

**Pull request closed without merging**